packaging/snapd.mk

```diff
@@ -0,0 +1,190 @@
+# Those knobs influence the build and installation process. The knobs cannot be
+# undefined if they are actually used by the Makefile below.
+
+# Set to 1 to build snapd with test key built in.
+with_test_keys?=$(error with_test_keys is unset, must be 1 or 0)
+# Set to 1 to build snapd with apparmor support.
+with_apparmor?=$(error with_apparmor is unset, must be 1 or 0)
+# Set to 1 to build snapd with things needed for the core/snapd snap.
+with_core_bits?=$(error with_core_bits is unset, must be 1 or 0)
+# Set to 1 to build snapd with alternate snap mount directory.
+with_alt_snap_mount_dir=$(error with_alt_snap_mount_dir is unset, must be 1 or 0)
+
+# ------------------------------------------------
+# There are no more control knobs after this point
+# ------------------------------------------------
+
+# Import path of snapd.
+import_path=github.com/snapcore/snapd
+
+# Set a set of variables to some well-known directories. If rpm is installed
+# then it is used as canonical reference. In absence of rpm the values are
+# built out of Debian-ish defaults.
+ifneq ($(shell which rpm),)
+prefix:=$(shell rpm -E %{_prefix})
+bindir:=$(shell rpm -E %{_bindir})
+sbindir:=$(shell rpm -E %{_sbindir})
+libexecdir:=$(shell rpm -E %{_libexecdir})
+mandir:=$(shell rpm -E %{_mandir})
+datadir:=$(shell rpm -E %{_datadir})
+localstatedir:=$(shell rpm -E %{_localstatedir})
+unitdir=$(shell rpm -E %{_unitdir})
+else
+$(error expected RPM to be used)
+prefix=/usr
+bindir=$(prefix)/bin
+sbindir=$(prefix)/sbin
+libexecdir=$(prefix)/lib
+mandir=$(prefix)/
+datadir=$(prefix)/share
+localstatedir=/var
+unitdir=$(prefix)/lib/systemd/system
+endif
+
+# Ensure that none of the directory variables are empty.
+vars=bindir sbindir libexecdir mandir datadir localstatedir unitdir
+$(foreach var,$(vars),$(if $(value $(var)),,$(error $(var) is empty or unset)))
+
+# This is usually set by %make_install. It is defined here to avoid warnings or
+# errors from referencing undefined variables.
+DESTDIR?=
+
+# Decide which of the two snap mount directories to use. This is not
+# referencing localstatedir because the code copes with only those two values
+# explicitly.
+ifeq ($(with_alt_snap_mount_dir),1)
+snap_mount_dir=/var/lib/snapd/snap
+else
+snap_mount_dir=/snap
+endif
+
+# The list of go binaries we are expected to build.
+go_binaries=snap snapctl snap-seccomp snap-update-ns snap-exec snapd
+
+# NOTE: This *depends* on building out of tree. Some of the built binaries
+# conflict with directory names in the tree.
+.PHONY: all
+all: $(go_binaries) 
+
+snap snapctl snap-seccomp:
+	go build -buildmode=pie $(import_path)/cmd/$@
+
+# Those two need to be built as static binaries. They run on the inside of a
+# nearly-arbitrary mount namespace that does not contain anything we can depend
+# on (no standard library, for example).
+snap-update-ns snap-exec:
+	go build -buildmode=default -ldflags '-extldflags "-static"' $(import_path)/cmd/$@
```

Comment:
At some point we'll also need to add `-tags osusergo` to use pure go implementation of `os/user` to avoid some `getpwent()` calls. I also have some changes for `osutil` that do the same trick.

Reply:
Sounds good.

Reply:
I think this can happen in a follow-up.